applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp

```diff
@@ -1,982 +1,966 @@
+  const double A = this->GetProperties()[CROSS_AREA];
+  const double L = this->CalculateReferenceLength();
+  const double rho = this->GetProperties()[DENSITY];
+
+  double total_mass = A * L * rho;
+  BoundedVector<double, msDimension> body_forces_node =
+      ZeroVector(msDimension);
+  BoundedVector<double, msLocalSize> body_forces_global =
+      ZeroVector(msLocalSize);
+
+  // assemble global Vector
+  for (int i = 0; i < msNumberOfNodes; ++i) {
+    body_forces_node =
+        total_mass *
+        this->GetGeometry()[i].FastGetSolutionStepValue(VOLUME_ACCELERATION) *
+        Ncontainer(0, i);
+
+    for (unsigned int j = 0; j < msDimension; ++j) {
+      body_forces_global[(i * msDimension) + j] = body_forces_node[j];
+    }
+  }
+
+  return body_forces_global;
+  KRATOS_CATCH("")
+}
+
+void TrussElement3D2N::GetValuesVector(Vector &rValues, int Step) {
+
+  KRATOS_TRY
+  if (rValues.size() != msLocalSize)
+    rValues.resize(msLocalSize, false);
+
+  for (int i = 0; i < msNumberOfNodes; ++i) {
+    int index = i * msDimension;
+    const auto &disp =
+        this->GetGeometry()[i].FastGetSolutionStepValue(DISPLACEMENT, Step);
+
+    rValues[index] = disp[0];
+    rValues[index + 1] = disp[1];
+    rValues[index + 2] = disp[2];
+  }
+  KRATOS_CATCH("")
+}
+
+void TrussElement3D2N::GetFirstDerivativesVector(Vector &rValues, int Step) {
+
+  KRATOS_TRY
+  if (rValues.size() != msLocalSize)
+    rValues.resize(msLocalSize, false);
+
+  for (int i = 0; i < msNumberOfNodes; ++i) {
+    int index = i * msDimension;
+    const auto &vel =
+        this->GetGeometry()[i].FastGetSolutionStepValue(VELOCITY, Step);
+
+    rValues[index] = vel[0];
+    rValues[index + 1] = vel[1];
+    rValues[index + 2] = vel[2];
+  }
+  KRATOS_CATCH("")
+}
+
+void TrussElement3D2N::GetSecondDerivativesVector(Vector &rValues, int Step) {
+
+  KRATOS_TRY
+  if (rValues.size() != msLocalSize)
+    rValues.resize(msLocalSize, false);
+
+  for (int i = 0; i < msNumberOfNodes; ++i) {
+    int index = i * msDimension;
+    const auto &acc =
+        this->GetGeometry()[i].FastGetSolutionStepValue(ACCELERATION, Step);
+
+    rValues[index] = acc[0];
+    rValues[index + 1] = acc[1];
+    rValues[index + 2] = acc[2];
+  }
+
+  KRATOS_CATCH("")
+}
+
+void TrussElement3D2N::CalculateLocalSystem(MatrixType &rLeftHandSideMatrix,
+                                            VectorType &rRightHandSideVector,
+                                            ProcessInfo &rCurrentProcessInfo) {
+
+  KRATOS_TRY
+  // calculate internal forces
+  BoundedVector<double, msLocalSize> internal_forces = ZeroVector(msLocalSize);
+  this->UpdateInternalForces(internal_forces);
+  // creating LHS
+  rLeftHandSideMatrix =
+      this->CreateElementStiffnessMatrix(rCurrentProcessInfo);
+  // create+compute RHS
+  rRightHandSideVector = -internal_forces;
+  // add bodyforces
+  if (this->CheckSelfWeight()) noalias(rRightHandSideVector) += this->CalculateBodyForces();
```

Comment:
the name is ambiguous, pls use sth like `HasSelfWeight`

Reply:
I haven't implemented that

Reply:
@KlausBSautter

Reply:
@philbucher  :whale: